File: watcher-go/ccpp.go
package main

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// Ports ccpp.py's CitizenCompassPacket: scans the project, cross-references
// ships/data/docs, and computes the same health-score weighting.

type shipInfo struct {
	Slug             string `json:"slug"`
	Path             string `json:"path"`
	Files            []string `json:"files"`
	HardpointsCount  int    `json:"hardpoints_count"`
	ViewerComplete   bool   `json:"viewer_complete"`
	ModelSize        int64  `json:"model_size,omitempty"`
}

type dataLayerInfo struct {
	Path         string         `json:"path"`
	FileCount    int            `json:"file_count"`
	TotalSizeMB  float64        `json:"total_size_mb"`
	FileTypes    map[string]int `json:"file_types"`
	SampleFiles  []string       `json:"sample_files"`
}

type ccppInventory struct {
	Ships      map[string]*shipInfo     `json:"ships"`
	DataLayers map[string]*dataLayerInfo `json:"data_layers"`
	Scripts    []map[string]interface{} `json:"scripts"`
	Models     []map[string]interface{} `json:"models"`
	Docs       []map[string]interface{} `json:"docs"`
}

type ccppCrossref struct {
	ShipsWithViewers    int            `json:"ships_with_viewers"`
	ShipsTotal          int            `json:"ships_total"`
	ViewersProgressPct  float64        `json:"viewers_progress_pct"`
	DataFilesByCategory map[string]int `json:"data_files_by_category"`
}

type ccppScores struct {
	DataCompleteness float64 `json:"data_completeness"`
	ViewerProgress   float64 `json:"viewer_progress"`
	Documentation    float64 `json:"documentation"`
	OverallHealth    float64 `json:"overall_health"`
}

type ccppPacket struct {
	Metadata  map[string]interface{} `json:"metadata"`
	Inventory ccppInventory          `json:"inventory"`
	Crossref  ccppCrossref           `json:"crossref"`
	Scores    ccppScores             `json:"scores"`
}

func scanProject() *ccppPacket {
	p := &ccppPacket{
		Metadata: map[string]interface{}{
			"format":  "CCPP-1.0",
			"project": "Citizen Compass",
			"created": time.Now().Format(time.RFC3339),
			"updated": time.Now().Format(time.RFC3339),
		},
		Inventory: ccppInventory{
			Ships:      map[string]*shipInfo{},
			DataLayers: map[string]*dataLayerInfo{},
		},
	}

	scanShips(p)
	scanDataLayer(p)
	scanScripts(p)
	scanModels(p)
	scanDocs(p)
	buildCrossref(p)
	calculateScores(p)

	p.Metadata["updated"] = time.Now().Format(time.RFC3339)
	p.Metadata["project_path"] = projectRoot
	p.Metadata["scan_complete"] = true

	return p
}

func scanShips(p *ccppPacket) {
	entries, err := os.ReadDir(shipsDir)
	if err != nil {
		return
	}
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		slug := e.Name()
		shipDir := filepath.Join(shipsDir, slug)
		info := &shipInfo{Slug: slug, Path: shipDir, Files: []string{}}

		modelPath := filepath.Join(shipDir, "model.glb")
		indexPath := filepath.Join(shipDir, "index.html")
		hpPath := filepath.Join(shipDir, "hardpoints.json")

		if st, err := os.Stat(modelPath); err == nil {
			info.Files = append(info.Files, "model.glb")
			info.ModelSize = st.Size()
		}
		if _, err := os.Stat(indexPath); err == nil {
			info.Files = append(info.Files, "index.html")
		}
		if _, err := os.Stat(hpPath); err == nil {
			info.Files = append(info.Files, "hardpoints.json")
			if raw, err := os.ReadFile(hpPath); err == nil {
				var data map[string]interface{}
				if json.Unmarshal(raw, &data) == nil {
					if hp, ok := data["hardpoints"].([]interface{}); ok {
						info.HardpointsCount = len(hp)
					}
					info.ViewerComplete = len(info.Files) == 3
				}
			}
		}

		if len(info.Files) > 0 {
			p.Inventory.Ships[slug] = info
		}
	}
}

func scanDataLayer(p *ccppPacket) {
	nested := filepath.Join(projectRoot, "data-layer")
	flat := filepath.Join(projectRoot, "data-layerrawhardpoints")

	var dataDir string
	if dirExists(nested) {
		dataDir = nested
	} else if dirExists(flat) {
		dataDir = flat
	}

	if dataDir == "" {
		matches, _ := filepath.Glob(filepath.Join(projectRoot, "data-layer*"))
		for _, m := range matches {
			if dirExists(m) {
				catalogDataFolder(p, m, filepath.Base(m))
			}
		}
		return
	}

	catalogDataFolder(p, dataDir, "data-layer")
}

func catalogDataFolder(p *ccppPacket, folder string, name string) {
	if !dirExists(folder) {
		return
	}
	fileCount := 0
	var totalSize int64
	fileTypes := map[string]int{}
	var sampleFiles []string

	filepath.Walk(folder, func(path string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() {
			return nil
		}
		fileCount++
		totalSize += info.Size()
		ext := strings.ToLower(filepath.Ext(path))
		fileTypes[ext]++
		if len(sampleFiles) < 3 {
			sampleFiles = append(sampleFiles, filepath.Base(path))
		}
		return nil
	})

	p.Inventory.DataLayers[name] = &dataLayerInfo{
		Path:        folder,
		FileCount:   fileCount,
		TotalSizeMB: round2(float64(totalSize) / (1024 * 1024)),
		FileTypes:   fileTypes,
		SampleFiles: sampleFiles,
	}
}

func scanScripts(p *ccppPacket) {
	matches, _ := filepath.Glob(filepath.Join(projectRoot, "*.py"))
	for _, m := range matches {
		st, err := os.Stat(m)
		if err != nil {
			continue
		}
		p.Inventory.Scripts = append(p.Inventory.Scripts, map[string]interface{}{
			"name": filepath.Base(m), "path": m, "size": st.Size(),
		})
	}
}

func scanModels(p *ccppPacket) {
	filepath.Walk(projectRoot, func(path string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() {
			return nil
		}
		ext := strings.ToLower(filepath.Ext(path))
		if ext == ".blend" || ext == ".glb" {
			p.Inventory.Models = append(p.Inventory.Models, map[string]interface{}{
				"name": filepath.Base(path), "path": path, "size_mb": round2(float64(info.Size()) / (1024 * 1024)),
			})
		}
		return nil
	})
}

func scanDocs(p *ccppPacket) {
	filepath.Walk(projectRoot, func(path string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() {
			return nil
		}
		if strings.ToLower(filepath.Ext(path)) == ".md" {
			p.Inventory.Docs = append(p.Inventory.Docs, map[string]interface{}{
				"name": filepath.Base(path), "path": path,
			})
		}
		return nil
	})
}

func buildCrossref(p *ccppPacket) {
	withViewers := 0
	for _, s := range p.Inventory.Ships {
		if s.ViewerComplete {
			withViewers++
		}
	}
	total := len(p.Inventory.Ships)
	p.Crossref = ccppCrossref{
		ShipsWithViewers:    withViewers,
		ShipsTotal:          total,
		DataFilesByCategory: map[string]int{},
	}
	if total > 0 {
		p.Crossref.ViewersProgressPct = round1(float64(withViewers) / float64(total) * 100)
	}
	for name, data := range p.Inventory.DataLayers {
		lower := strings.ToLower(name)
		switch {
		case strings.Contains(lower, "raw"):
			p.Crossref.DataFilesByCategory["raw"] = data.FileCount
		case strings.Contains(lower, "processed"):
			p.Crossref.DataFilesByCategory["processed"] = data.FileCount
		case strings.Contains(lower, "export"):
			p.Crossref.DataFilesByCategory["exports"] = data.FileCount
		}
	}
}

func calculateScores(p *ccppPacket) {
	dataFiles := 0
	for name, data := range p.Inventory.DataLayers {
		if strings.Contains(strings.ToLower(name), "raw") {
			dataFiles = data.FileCount
			break
		}
	}
	p.Scores.DataCompleteness = minF(100, round1(float64(dataFiles)/232*100))
	p.Scores.ViewerProgress = p.Crossref.ViewersProgressPct
	p.Scores.Documentation = minF(100, float64(len(p.Inventory.Docs)*20))
	p.Scores.OverallHealth = round1(
		p.Scores.DataCompleteness*0.40 + p.Scores.ViewerProgress*0.50 + p.Scores.Documentation*0.10,
	)
}

func saveCcpp(p *ccppPacket, path string) error {
	encoded, err := json.MarshalIndent(p, "", "  ")
	if err != nil {
		return err
	}
	checksum := fmt.Sprintf("%x", sha256.Sum256(encoded))[:16]
	p.Metadata["checksum"] = checksum
	encoded, err = json.MarshalIndent(p, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, encoded, 0644)
}

func loadCcpp(path string) (*ccppPacket, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var p ccppPacket
	if err := json.Unmarshal(raw, &p); err != nil {
		return nil, err
	}
	return &p, nil
}

// rescanAndScore mirrors inbox_watcher.py's rescan_and_score(): re-scan the
// whole project, save citizen-compass.ccpp, log the refreshed health score.
func rescanAndScore() {
	p := scanProject()
	if err := saveCcpp(p, ccppFile); err != nil {
		logMsg("could not save %s: %v", ccppFile, err)
		return
	}
	logMsg("Health score refreshed: %.1f/100 (data %.1f%%, viewers %.1f%%, docs %.1f%%)",
		p.Scores.OverallHealth, p.Scores.DataCompleteness, p.Scores.ViewerProgress, p.Scores.Documentation)
}

func round1(v float64) float64 {
	return float64(int(v*10+0.5)) / 10
}
func round2(v float64) float64 {
	return float64(int(v*100+0.5)) / 100
}
func minF(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

File: ccpp.py
"""
Citizen Compass Project Packet (CCPP)

A self-contained project archive application that:
- Scans project folders for data files
- Auto-detects file types and content
- Organizes by category (ships, data, configs, docs)
- Validates completeness and cross-references
- Scores health/progress
- Compacts into a single portable JSON file
- Can be passed between AIs and systems

Usage:
    python ccpp.py create <project_path>       # Scan and create packet
    python ccpp.py update <packet.ccpp>        # Re-scan and update
    python ccpp.py inspect <packet.ccpp>       # View contents
    python ccpp.py extract <packet.ccpp>       # Unpack files
    python ccpp.py validate <packet.ccpp>      # Check integrity

Examples:
    python ccpp.py create C:\\Users\\david\\citizen-compass
    python ccpp.py inspect citizen-compass.ccpp
    python ccpp.py validate citizen-compass.ccpp
"""

import json
import sys
import hashlib
from pathlib import Path
from datetime import datetime
from collections import defaultdict


class CitizenCompassPacket:
    """Self-contained project packet with auto-detection and scoring."""

    def __init__(self):
        self.metadata = {
            "format": "CCPP-1.0",
            "project": "Citizen Compass",
            "created": datetime.now().isoformat(),
            "updated": datetime.now().isoformat(),
        }
        self.inventory = {
            "ships": {},
            "data_layers": {},
            "scripts": [],
            "models": [],
            "configs": [],
            "docs": [],
        }
        self.crossref = {}
        self.scores = {
            "data_completeness": 0,
            "viewer_progress": 0,
            "documentation": 0,
            "overall_health": 0,
        }
        self.file_index = {}

    def scan_project(self, project_path):
        """Scan entire project and auto-detect files."""
        project_path = Path(project_path)
        if not project_path.exists():
            print(f"❌ Path not found: {project_path}")
            return False

        print(f"📁 Scanning: {project_path}")

        # Scan key directories
        self._scan_ships(project_path)
        self._scan_data_layer(project_path)
        self._scan_scripts(project_path)
        self._scan_models(project_path)
        self._scan_docs(project_path)

        # Cross-reference and score
        self._build_crossref()
        self._calculate_scores()

        self.metadata["updated"] = datetime.now().isoformat()
        self.metadata["project_path"] = str(project_path)
        self.metadata["scan_complete"] = True

        return True

    def _scan_ships(self, project_path):
        """Scan ship viewer folders."""
        ships_dir = project_path / "tests" / "testing-site" / "ships"
        if not ships_dir.exists():
            return

        for ship_folder in ships_dir.glob("*/"):
            slug = ship_folder.name
            ship_data = {
                "slug": slug,
                "path": str(ship_folder),
                "files": [],
                "hardpoints_count": 0,
                "viewer_complete": False,
            }

            # Check for key files
            model = ship_folder / "model.glb"
            index = ship_folder / "index.html"
            hardpoints = ship_folder / "hardpoints.json"

            if model.exists():
                ship_data["files"].append("model.glb")
                ship_data["model_size"] = model.stat().st_size
            if index.exists():
                ship_data["files"].append("index.html")
            if hardpoints.exists():
                ship_data["files"].append("hardpoints.json")
                try:
                    with open(hardpoints, "r", encoding="utf-8") as f:
                        hp_data = json.load(f)
                        ship_data["hardpoints_count"] = len(
                            hp_data.get("hardpoints", [])
                        )
                        ship_data["viewer_complete"] = (
                            len(ship_data["files"]) == 3
                        )
                except:
                    pass

            if ship_data["files"]:
                self.inventory["ships"][slug] = ship_data
                self.file_index[str(ship_folder)] = {
                    "type": "ship",
                    "slug": slug,
                }

    def _scan_data_layer(self, project_path):
        """Scan data-layer folders."""
        data_dir = project_path / "data-layer" if (project_path / "data-layer").exists() else \
                  project_path / "data-layerrawhardpoints" if (project_path / "data-layerrawhardpoints").exists() else None

        if not data_dir or not data_dir.exists():
            # Try checking for individual data-layer* folders
            for item in project_path.glob("data-layer*"):
                if item.is_dir():
                    self._catalog_data_folder(item, item.name)
            return

        self._catalog_data_folder(data_dir, "data-layer")

    def _catalog_data_folder(self, folder, name):
        """Catalog a data folder and its contents."""
        if not folder.exists():
            return

        file_count = 0
        total_size = 0
        file_types = defaultdict(int)
        sample_files = []

        for f in folder.rglob("*"):
            if f.is_file():
                file_count += 1
                total_size += f.stat().st_size
                ext = f.suffix.lower()
                file_types[ext] += 1
                if len(sample_files) < 3:
                    sample_files.append(f.name)

        self.inventory["data_layers"][name] = {
            "path": str(folder),
            "file_count": file_count,
            "total_size_mb": round(total_size / (1024 * 1024), 2),
            "file_types": dict(file_types),
            "sample_files": sample_files,
        }

    def _scan_scripts(self, project_path):
        """Scan for Python/JS scripts."""
        for py_file in project_path.glob("*.py"):
            self.inventory["scripts"].append({
                "name": py_file.name,
                "path": str(py_file),
                "size": py_file.stat().st_size,
            })

    def _scan_models(self, project_path):
        """Scan for 3D model files."""
        for blend in project_path.rglob("*.blend"):
            self.inventory["models"].append({
                "name": blend.name,
                "path": str(blend),
                "size_mb": round(blend.stat().st_size / (1024 * 1024), 2),
            })
        for glb in project_path.rglob("*.glb"):
            self.inventory["models"].append({
                "name": glb.name,
                "path": str(glb),
                "size_mb": round(glb.stat().st_size / (1024 * 1024), 2),
            })

    def _scan_docs(self, project_path):
        """Scan for documentation."""
        for doc in project_path.rglob("*.md"):
            self.inventory["docs"].append({
                "name": doc.name,
                "path": str(doc),
            })

    def _build_crossref(self):
        """Cross-reference data (ships ↔ data files ↔ viewers)."""
        self.crossref = {
            "ships_with_viewers": len(
                [s for s in self.inventory["ships"].values() if s["viewer_complete"]]
            ),
            "ships_total": len(self.inventory["ships"]),
            "viewers_progress_pct": 0,
            "data_files_by_category": {},
        }

        if self.crossref["ships_total"] > 0:
            self.crossref["viewers_progress_pct"] = round(
                (self.crossref["ships_with_viewers"] / self.crossref["ships_total"]) * 100, 1
            )

        # Categorize data files
        for name, data in self.inventory["data_layers"].items():
            if "raw" in name.lower():
                self.crossref["data_files_by_category"]["raw"] = data["file_count"]
            elif "processed" in name.lower():
                self.crossref["data_files_by_category"]["processed"] = data["file_count"]
            elif "export" in name.lower():
                self.crossref["data_files_by_category"]["exports"] = data["file_count"]

    def _calculate_scores(self):
        """Calculate health/progress scores (0-100)."""
        # Data completeness: based on weapon hardpoint files
        raw_weapons = self.inventory["data_layers"].get(
            [k for k in self.inventory["data_layers"] if "raw" in k.lower()][0]
            if any("raw" in k.lower() for k in self.inventory["data_layers"])
            else "none",
            {}
        )
        data_files = raw_weapons.get("file_count", 0)
        self.scores["data_completeness"] = min(100, round((data_files / 232) * 100, 1))

        # Viewer progress: ships with complete viewers / target
        self.scores["viewer_progress"] = self.crossref.get(
            "viewers_progress_pct", 0
        )

        # Documentation: number of doc files
        self.scores["documentation"] = min(
            100, len(self.inventory["docs"]) * 20
        )

        # Overall: weighted average
        self.scores["overall_health"] = round(
            (
                self.scores["data_completeness"] * 0.40
                + self.scores["viewer_progress"] * 0.50
                + self.scores["documentation"] * 0.10
            ),
            1,
        )

    def save(self, filename):
        """Save packet to file."""
        packet = {
            "metadata": self.metadata,
            "inventory": self.inventory,
            "crossref": self.crossref,
            "scores": self.scores,
        }

        # Compute checksum
        packet_json = json.dumps(packet, indent=2).encode()
        packet["metadata"]["checksum"] = hashlib.sha256(packet_json).hexdigest()[:16]

        with open(filename, "w", encoding="utf-8") as f:
            json.dump(packet, f, indent=2)

        print(f"✅ Packet saved: {filename}")
        return filename

    def load(self, filename):
        """Load packet from file."""
        # JSON is UTF-8 by spec (RFC 8259). Without an explicit encoding these
        # fall back to the Windows ANSI codepage (cp1252), which cannot decode
        # non-ASCII ship names - e.g. the Xi'an "...an'tok.yaai" path raised
        # UnicodeDecodeError on byte 0x81 and took the whole handoff
        # regeneration down with it. The file was always valid; the reader was not.
        with open(filename, "r", encoding="utf-8") as f:
            packet = json.load(f)

        self.metadata = packet.get("metadata", {})
        self.inventory = packet.get("inventory", {})
        self.crossref = packet.get("crossref", {})
        self.scores = packet.get("scores", {})

        return True

    def inspect(self):
        """Print readable summary of packet contents."""
        print("\n" + "=" * 70)
        print("CITIZEN COMPASS PROJECT PACKET")
        print("=" * 70)

        print(f"\n📊 PROJECT HEALTH SCORE: {self.scores.get('overall_health', 0)}/100")
        print("\nBreakdown:")
        print(f"  • Data Completeness: {self.scores.get('data_completeness', 0)}%")
        print(f"  • Viewer Progress: {self.scores.get('viewer_progress', 0)}%")
        print(f"  • Documentation: {self.scores.get('documentation', 0)}%")

        print("\n📈 INVENTORY SUMMARY:")
        print(f"  • Ships (total): {self.crossref.get('ships_total', 0)}")
        print(f"  • Ships (viewers complete): {self.crossref.get('ships_with_viewers', 0)}")
        print(f"  • Viewer Progress: {self.crossref.get('viewers_progress_pct', 0)}%")
        print(f"  • Data Files: {sum(d.get('file_count', 0) for d in self.inventory['data_layers'].values())}")
        print(f"  • Scripts: {len(self.inventory['scripts'])}")
        print(f"  • 3D Models: {len(self.inventory['models'])}")
        print(f"  • Documentation: {len(self.inventory['docs'])}")

        print("\n🚢 SHIP VIEWERS STATUS:")
        complete = [s for s in self.inventory["ships"].values() if s.get("viewer_complete")]
        for ship in complete[:5]:
            print(f"  ✅ {ship['slug']}: {ship['hardpoints_count']} hardpoints")
        if len(complete) > 5:
            print(f"  ... and {len(complete) - 5} more")

        if self.inventory["ships"]:
            incomplete = [s for s in self.inventory["ships"].values() if not s.get("viewer_complete")]
            if incomplete:
                print(f"\n  ⏳ In Progress: {len(incomplete)}")
                for ship in incomplete[:3]:
                    print(f"     • {ship['slug']}: {len(ship['files'])} files")

        print("\n📁 DATA LAYERS:")
        for name, data in self.inventory["data_layers"].items():
            print(f"  • {name}: {data['file_count']} files ({data['total_size_mb']} MB)")

        print("\n" + "=" * 70)


def main():
    if len(sys.argv) < 2:
        print(__doc__)
        return

    command = sys.argv[1]
    packet = CitizenCompassPacket()

    if command == "create" and len(sys.argv) >= 3:
        project_path = sys.argv[2]
        output_file = sys.argv[3] if len(sys.argv) > 3 else "citizen-compass.ccpp"

        packet.scan_project(project_path)
        packet.save(output_file)
        packet.inspect()

    elif command == "update" and len(sys.argv) >= 3:
        ccpp_file = sys.argv[2]
        packet.load(ccpp_file)
        project_path = packet.metadata.get("project_path")
        if project_path:
            packet.scan_project(project_path)
            packet.save(ccpp_file)
            packet.inspect()
        else:
            print("❌ No project path in packet")

    elif command == "inspect" and len(sys.argv) >= 3:
        ccpp_file = sys.argv[2]
        packet.load(ccpp_file)
        packet.inspect()

    elif command == "validate" and len(sys.argv) >= 3:
        ccpp_file = sys.argv[2]
        packet.load(ccpp_file)
        checksum = packet.metadata.get("checksum", "")
        print(f"✅ Packet valid (checksum: {checksum})")
        print(f"   Created: {packet.metadata.get('created')}")
        print(f"   Updated: {packet.metadata.get('updated')}")

    else:
        print(__doc__)


if __name__ == "__main__":
    main()
